src/decode/ykpt/packet_parser/mod.rs

```diff
@@ -0,0 +1,303 @@
+//! A packet parser for the Yk PT trace decoder.
+
+use crate::errors::HWTracerError;
+use deku::{bitvec::BitSlice, DekuRead};
+use std::{fmt::Write, iter::Iterator};
+
+mod packets;
+use packets::*;
+
+#[derive(Clone, Copy, Debug)]
+enum PacketParserState {
+    /// Initial state, waiting for a PSB packet.
+    Init,
+    /// The "normal" decoding state.
+    Normal,
+    /// We are decoding a PSB+ sequence.
+    PSBPlus,
+}
+
+impl PacketParserState {
+    /// Returns the kinds of packet that are valid for the state.
+    fn valid_packets(&self) -> &'static [PacketKind] {
+        // Note that the parser will attempt to match packet kinds in the order that they appear in
+        // the returned slice. For best performance, the returned slice should be sorted, most
+        // frequently expected packet kinds first.
+        //
+        // OPT: The order below is a rough guess based on what limited traces I've seen. Benchmark
+        // and optimise.
+        match self {
+            Self::Init => &[PacketKind::PSB],
+            Self::Normal => &[
+                PacketKind::ShortTNT,
+                PacketKind::PAD,
+                PacketKind::FUP,
+                PacketKind::TIP,
+                PacketKind::CYC,
+                PacketKind::LongTNT,
+                PacketKind::PSB,
+                PacketKind::MODE,
+                PacketKind::TIPPGE,
+                PacketKind::TIPPGD,
+            ],
+            Self::PSBPlus => &[PacketKind::CBR, PacketKind::PSBEND],
+        }
+    }
+
+    /// Check if we need to transition state after parsing a given packet kind.
```

Comment:
I can't parse this.

Reply:
Parses fine here, but we can expand it to something more explicit:

> Check if the parser needs to transition to a new state as a result of parsing a certain kind of packet.

Reply:
Your rephrasing is much better!

Reply:
Fixed in [3567abf](https://github.com/ykjit/hwtracer/pull/105/commits/3567abfb722b6106c105a07bc9a0b4ceda47d3a4)